### DISCUSSION TITLE
High frequencies with small changes possible?

### DISCUSSION THREAD

**Discussion by peterbmarks** [khoih-prog/RP2040_PWM#14] [Q&A] [answered]
Thanks for this library. I'm interested in generating radio frequencies such as 7Mhz. I find that the available frequencies generated with your library are in quite big steps.

My application is to generate [WSPR](https://en.wikipedia.org/wiki/WSPR_(amateur_radio_software)) which consists of an RF frequency that slowly moves between frequencies in 1.4648 Hz increments.

Is the accuracy of the frequency a limitation of the PIO on the 2040? or is there a way I could use the library to achieve this?

**Comment by khoih-prog** [comment, marked as answer]
Hi @peterbmarks 

> Is the accuracy of the frequency a limitation of the PIO on the 2040?

That sounds interesting project, but your concern, IMHO, is more accurate than what the RP2040 can provide.

The TOP register to control PWM is only 16-bit, giving 64K resolution can't provide the 1.4648 Hz increments in MHz range (nearly 32-bit resolution)

I suggest you use the hardware solution, such as **Si570 programmable crystal oscillator**, as described in [**Standalone Multiband WSPR
beacon transmitter (MEPT)**](http://www.wsprnet.org/drupal/sites/wsprnet.org/files/si570wspr.pdf)

> The following is a design of a standalone MEPT beacon used for WSPR broadcasts [2]. It is
> based on a Si570 programmable crystal oscillator (silabs.com) which is controlled by means of an
> AVR attiny45 microcontroller. The Si570 device is FSK modulated by the AVR via a sequence
> of I2C control commands that are sent to slightly change the frequency for every transmitted
> WSPR symbol. The Si570 device outputs a power of 10mW which is sufficient for WSPR
> operation (i.e. an antenna with some LPF can be directly attached). For increased power up to
> 500mW, a small BS170 based power amplifier may be added. As the beacon alternately changes
> frequencies, the filters should be made switchable. Optionally this filter switching could be
> controlled by the microcontroller via I2C and a PCF8574 device (not supported by supplied
> firmware).

Just use the **RP2040** to replace the old **AVR attiny45**

**Reply by peterbmarks** [reply]
Thanks @khoih-prog for getting back to me. Yes, I can see the issue. I have built these transmitters in the past using the Si5351 clock generator but was hoping to reduce the parts count.
Thanks again for your contributions.
Peter
